### PR TITLE
FlatField refactor

### PR DESCRIPTION
**Pull request by profxj**
Introduces FlatField class to guide flat fielding and continues
the refactoring of calibration steps.  Here are the specific changes:

- Introduces ```FlatField``` class
- Adds FlatField Notebook, tests
- Pushes flat field algorithms into ```core/arflat.py```
- Main flatfield method broken into a few pieces
- Further refactoring of ```armasters.py```
- Further refactoring related to settings and ```ScienceExposure```

### CODE REVIEW

**Comment by profxj** [comment]
This Notebook shows some of the functionality:

https://github.com/PYPIT/PYPIT/blob/flatfield_refactor/doc/nb/FlatField.ipynb

**Comment by jhennawi** [comment]
The flat field image has crazy values where there is no slit data. I would fix this by multiplying it by (slit mask > 0.0) where slitmask is a 2-d image which has the slit number for all the pixels belonging to a slit.

**Comment by jhennawi** [comment]
It does not appear that the slit_profile is being divided out of these flats, although it is in the source code. I invite @profxj or @rcooke-ast to comment. We will definitely need illumination flats for the extraction.

File: pypit/flatfield.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+        return modvals, nrmvals, msblaze_slit, blazeext_slit, iextrap_slit
+
+    def run(self, datasec_img, armed=False):
+        """
+        Main driver to generate normalized flat field
+
+        Code flow:
+          1.  Generate the pixelflat image (if necessary)
+          2.  Apply the gain
+          3.  Prepare b-spline knot spacing
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypit/flatfield.py]: have added a TODO in slit_profile

File: pypit/flatfield.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+        self.slit_profiles
+
+        """
+
+        # Build the pixel flat (as needed)
+        if self.mspixelflat is None:
+            self.mspixelflat = self.build_pixflat()
+
+        # Apply gain
+        self.apply_gain(datasec_img)
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypit/flatfield.py]: agreed.  This was updated in a future branch but
I've moved it forward now.

File: pypit/flatfield.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+        # Return
+        return modvals, nrmvals, msblaze_slit, blazeext_slit, iextrap_slit
+
+    def run(self, datasec_img, armed=False):
+        """
+        Main driver to generate normalized flat field
+
+        Code flow:
+          1.  Generate the pixelflat image (if necessary)
+          2.  Apply the gain
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypit/flatfield.py]: same

File: pypit/flatfield.py [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+        if armed:
+            if np.sum(self.extrap_slit) != 0.0:
+                slit_profiles, mstracenrm, msblaze = arflat.slit_profile_pca(
+                    self.mspixelflat, self.tilts, self.msblaze, self.extrap_slit, self.slit_profiles,
+                    self.slits_dict['lcen'], self.slits_dict['rcen'], self.slits_dict['pixwid'],
+                    self.slits_dict['slitpix'])
+
+        # Apply slit profile
+        winpp = np.where(self.slit_profiles != 0.0)
+        self.mspixelflatnrm[winpp] /= self.slit_profiles[winpp]
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypit/flatfield.py]: Started Issue #347

**Comment by profxj** [comment]
Have set pixels outside of the slits to 1.0

Have started Issue #347 on the last comment Joe made.

**Pull request merged**